<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Jason Willwerscheid" />


<title>Flashier benchmarking</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">FLASHvestigations</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/willwerscheid/FLASHvestigations">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Flashier benchmarking</h1>
<h4 class="author">Jason Willwerscheid</h4>
<h4 class="date">1/11/2019</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#report"> Report <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-07-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>FLASHvestigations/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.2.0). The <em>Report</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="report" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20180714code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20180714)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20180714code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20180714)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwillwerscheidFLASHvestigationstree1037f1105d5b886c1e9261ce821ebd23646a2a04targetblank1037f11a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/willwerscheid/FLASHvestigations/tree/1037f1105d5b886c1e9261ce821ebd23646a2a04" target="_blank">1037f11</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwillwerscheidFLASHvestigationstree1037f1105d5b886c1e9261ce821ebd23646a2a04targetblank1037f11a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    code/.DS_Store
    Ignored:    code/flashier_bench/.DS_Store
    Ignored:    data/flashier_bench/

Untracked files:
    Untracked:  code/ebspca.R
    Untracked:  code/flashier_bench/.test2.swp
    Untracked:  code/lowrank.R

Unstaged changes:
    Modified:   analysis/flashier_features.Rmd
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/willwerscheid/FLASHvestigations/blob/1037f1105d5b886c1e9261ce821ebd23646a2a04/analysis/flashier_bench.Rmd" target="_blank">1037f11</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-07-02
</td>
<td>
wflow_publish(“analysis/flashier_bench.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/willwerscheid/FLASHvestigations/64ed09f448d4bb34b38f6b6dbffb618a89a06d7f/docs/flashier_bench.html" target="_blank">64ed09f</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-01-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/willwerscheid/FLASHvestigations/blob/7951c3c5abaab0d3abf231ee2f892ca843e8efe0/analysis/flashier_bench.Rmd" target="_blank">7951c3c</a>
</td>
<td>
Jason Willwerscheid
</td>
<td>
2019-01-12
</td>
<td>
workflowr::wflow_publish(“analysis/flashier_bench.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="gtex-eqtl-data" class="section level2">
<h2>GTEx eQTL data</h2>
<p>First, I fit the “strong” subset of SNP-gene association statistics used in Urbut, Wang, Carbonetto, and Stephens (2018) (the <code>strong.z</code> dataset found <a href="https://stephenslab.github.io/gtexresults/gtexdata.html">here</a>). I fit five FLASH factors with scale-mixture-of-normal and point-normal priors using <code>flashr</code> and <code>flashier</code>.</p>
<p>The data is a dense 16k x 44 matrix that takes up 7.0 MB of memory when loaded into <code>R</code>. I used the <code>broadwl</code> partition of the <code>midway2</code> RCC cluster with 4 CPUs and 32 GB of memory, and I used Gao Wang’s <code>monitor_memory.py</code> script to test memory usage, as recommended in Peter Carbonetto’s large-scale data analysis <a href="https://github.com/pcarbo/R-survival-large-scale">tutorial</a>.</p>
<table>
<caption>GTEx eQTL data</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">VMS (GB)</th>
<th align="right">RSS (GB)</th>
<th align="right">Init (s/factor)</th>
<th align="right">Greedy (s/iter)</th>
<th align="right">Backfit (s/iter)</th>
<th align="right">Backfit iters</th>
<th align="right">ELBO</th>
<th align="right">Time (min)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>flashr.normal.mix</td>
<td align="right">0.62</td>
<td align="right">0.37</td>
<td align="right">0.96</td>
<td align="right">0.47</td>
<td align="right">0.43</td>
<td align="right">285</td>
<td align="right">-1324706</td>
<td align="right">3.25</td>
</tr>
<tr class="even">
<td>flashier.normal.mix</td>
<td align="right">0.48</td>
<td align="right">0.30</td>
<td align="right">0.13</td>
<td align="right">0.31</td>
<td align="right">0.25</td>
<td align="right">266</td>
<td align="right">-1324702</td>
<td align="right">1.94</td>
</tr>
<tr class="odd">
<td>flashr.point.normal</td>
<td align="right">0.57</td>
<td align="right">0.32</td>
<td align="right">0.97</td>
<td align="right">0.12</td>
<td align="right">0.10</td>
<td align="right">300</td>
<td align="right">-1326412</td>
<td align="right">0.87</td>
</tr>
<tr class="even">
<td>flashier.point.normal</td>
<td align="right">0.43</td>
<td align="right">0.24</td>
<td align="right">0.17</td>
<td align="right">0.09</td>
<td align="right">0.09</td>
<td align="right">263</td>
<td align="right">-1326413</td>
<td align="right">0.61</td>
</tr>
</tbody>
</table>
</div>
<div id="montoro-et-al.3-scrna-seq-data" class="section level2">
<h2>Montoro et al. 3’ scRNA-seq data</h2>
<p>Next, I fit the droplet-based 3’ scRNA-seq dataset analyzed in Montoro et al. (2018) (the data can be obtained <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103354">here</a>). I removed all genes with nonzero counts in five or fewer cells, performed a log-plus-one transform of the data, and fit five FLASH factors using scale-mixture-of-normal and point-normal priors.</p>
<p>The data matrix is 15k x 7k and takes up 838 MB of memory when loaded into <code>R</code> as a dense matrix. However, only 11.2% of entries are nonzero, so the data can also be loaded as a sparse <code>Matrix</code> object, in which case the data takes up 142 MB of memory (only <code>flashier</code> supports objects of class <code>Matrix</code>). All fits were performed on the <code>broadwl</code> partition of the <code>midway2</code> RCC cluster using 4 CPUs and 32 GB of memory.</p>
<table>
<caption>Montoro 3’ scRNA data</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">VMS (GB)</th>
<th align="right">RSS (GB)</th>
<th align="right">Init (s/factor)</th>
<th align="right">Greedy (s/iter)</th>
<th align="right">Backfit (s/iter)</th>
<th align="right">Backfit iters</th>
<th align="right">ELBO</th>
<th align="right">Time (min)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>flashr.normal.mix</td>
<td align="right">11.69</td>
<td align="right">11.44</td>
<td align="right">37.24</td>
<td align="right">15.14</td>
<td align="right">9.49</td>
<td align="right">315</td>
<td align="right">2354585</td>
<td align="right">56.72</td>
</tr>
<tr class="even">
<td>flashier.normal.mix.</td>
<td align="right">3.25</td>
<td align="right">3.07</td>
<td align="right">3.52</td>
<td align="right">1.18</td>
<td align="right">1.19</td>
<td align="right">293</td>
<td align="right">2354637</td>
<td align="right">6.76</td>
</tr>
<tr class="odd">
<td>flashier.sprs.normal.mix</td>
<td align="right">0.96</td>
<td align="right">0.78</td>
<td align="right">0.97</td>
<td align="right">0.45</td>
<td align="right">0.45</td>
<td align="right">294</td>
<td align="right">2354650</td>
<td align="right">2.51</td>
</tr>
<tr class="even">
<td>flashr.point.normal</td>
<td align="right">11.78</td>
<td align="right">11.07</td>
<td align="right">37.46</td>
<td align="right">13.28</td>
<td align="right">8.99</td>
<td align="right">105</td>
<td align="right">2322665</td>
<td align="right">23.94</td>
</tr>
<tr class="odd">
<td>flashier.point.normal</td>
<td align="right">3.17</td>
<td align="right">2.98</td>
<td align="right">3.48</td>
<td align="right">1.02</td>
<td align="right">1.00</td>
<td align="right">30</td>
<td align="right">2319044</td>
<td align="right">1.35</td>
</tr>
<tr class="even">
<td>flashier.sprs.point.normal</td>
<td align="right">0.87</td>
<td align="right">0.69</td>
<td align="right">0.97</td>
<td align="right">0.27</td>
<td align="right">0.26</td>
<td align="right">30</td>
<td align="right">2319044</td>
<td align="right">0.36</td>
</tr>
</tbody>
</table>
</div>
<div id="montoro-et-al.full-length-scrna-seq-data" class="section level2">
<h2>Montoro et al. full-length scRNA-seq data</h2>
<p>Finally, I fit the larger full-length scRNA “PulseSeq” dataset from Montoro et al. (2018). The dataset is about ten times larger than the droplet-based scRNA-seq dataset, so it was not feasible to use <code>flashr</code>. I again removed genes with nonzero counts in five or fewer cells, performed a log-plus-one transform of the data, and fit five factors using scale-mixture-of-normal and point-normal priors.</p>
<p>The dataset is 19k x 66k, with 10.6% of entries not equal to zero, and occupies 1.5 GB of memory when loaded into <code>R</code> as a sparse <code>Matrix</code> object. (I did not attempt to fit a larger <code>matrix</code> object.) The fit was again performed on <code>broadwl</code> using 4 CPUs and 32 GB of memory.</p>
<table>
<caption>Montoro full-length scRNA data</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">VMS (GB)</th>
<th align="right">RSS (GB)</th>
<th align="right">Init (s/factor)</th>
<th align="right">Greedy (s/iter)</th>
<th align="right">Backfit (s/iter)</th>
<th align="right">Backfit iters</th>
<th align="right">ELBO</th>
<th align="right">Time (min)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>flashier.sprs.normal.mix</td>
<td align="right">7.69</td>
<td align="right">7.54</td>
<td align="right">9.42</td>
<td align="right">2.78</td>
<td align="right">2.68</td>
<td align="right">20</td>
<td align="right">125716192</td>
<td align="right">4.33</td>
</tr>
<tr class="even">
<td>flashier.sprs.point.normal</td>
<td align="right">7.69</td>
<td align="right">7.54</td>
<td align="right">9.42</td>
<td align="right">2.18</td>
<td align="right">2.04</td>
<td align="right">8</td>
<td align="right">125635635</td>
<td align="right">3.16</td>
</tr>
</tbody>
</table>
</div>
<div id="code" class="section level2">
<h2>Code</h2>
The benchmarking code can be browsed <a href="https://github.com/willwerscheid/FLASHvestigations/tree/master/code/flashier_bench">here</a>. The <code>R</code> package versions used are those that appear in the session information below (“other attached packages”). <br> <br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.3 (2019-03-11)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.5

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mixsqp_0.1-119 ashr_2.2-38    ebnm_0.1-23    flashier_0.1.4
[5] flashr_0.6-6  

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.1        highr_0.8         plyr_1.8.4       
 [4] pillar_1.3.1      compiler_3.5.3    git2r_0.25.2     
 [7] workflowr_1.2.0   iterators_1.0.10  tools_3.5.3      
[10] digest_0.6.18     evaluate_0.13     tibble_2.1.1     
[13] gtable_0.3.0      lattice_0.20-38   pkgconfig_2.0.2  
[16] rlang_0.3.1       Matrix_1.2-15     foreach_1.4.4    
[19] yaml_2.2.0        parallel_3.5.3    xfun_0.6         
[22] dplyr_0.8.0.1     stringr_1.4.0     knitr_1.22       
[25] fs_1.2.7          tidyselect_0.2.5  rprojroot_1.3-2  
[28] grid_3.5.3        glue_1.3.1        R6_2.4.0         
[31] rmarkdown_1.12    reshape2_1.4.3    purrr_0.3.2      
[34] ggplot2_3.2.0     magrittr_1.5      whisker_0.3-2    
[37] backports_1.1.3   scales_1.0.0      codetools_0.2-16 
[40] htmltools_0.3.6   MASS_7.3-51.1     assertthat_0.2.1 
[43] softImpute_1.4    colorspace_1.4-1  stringi_1.4.3    
[46] lazyeval_0.2.2    doParallel_1.0.14 pscl_1.5.2       
[49] munsell_0.5.0     truncnorm_1.0-8   SQUAREM_2017.10-1
[52] crayon_1.3.4     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
