---
title: "Flashier benchmarking"
author: "Jason Willwerscheid"
date: "1/11/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(flashr)
library(flashier)
library(ebnm)
library(ashr)
library(mixsqp)
```

## GTEx eQTL data

First, I fit the "strong" subset of SNP-gene association statistics used in Urbut, Wang, Carbonetto, and Stephens (2018) (the `strong.z` dataset found [here](https://stephenslab.github.io/gtexresults/gtexdata.html)). I fit five FLASH factors with scale-mixture-of-normal and point-normal priors using `flashr` and  `flashier`.

The data is a dense 16k x 44 matrix that takes up 7.0 MB of memory when loaded into `R`. I used the `broadwl` partition of the `midway2` RCC cluster with 4 CPUs and 32 GB of memory, and I used Gao Wang's `monitor_memory.py` script to test memory usage, as recommended in Peter Carbonetto's large-scale data analysis [tutorial](https://github.com/pcarbo/R-survival-large-scale).

```{r res, echo = FALSE}
cols <- c("VMS (GB)",
          "RSS (GB)",
          "Init (s/factor)",
          "Greedy (s/iter)",
          "Backfit (s/iter)",
          "Backfit iters",
          "ELBO",
          "Time (min)")

res <- data.frame(
  mem.VMS = c(0.62, 0.48, 0.57, 0.43),
  mem.RSS = c(0.37, 0.30, 0.32, 0.24),
  time.per.init = c(7.14, 2.20, 5.46, 1.30) / 5,
  time.per.iter.greedy = c(64.3 / (3 + 6 + 46 + 50 + 31), 
                           47.8 / (1 + 11 + 39 + 67 + 35),
                           16.0 / (3 + 6 + 44 + 45 + 32),
                           12.6 / (1 + 11 + 37 + 62 + 36)),
  time.per.iter.bf = c(123.8 / (5*57),
                       66.5 / (5*43 + 4*(51 - 43) + 3*(53 - 51) + 2*(59 - 53) 
                               + 1*(60 - 59)),
                       30.8 / (5*60),
                       22.6 / (5*44 + 4*(46 - 44) + 3*(49 - 46) + 2*(62 - 49))),
  niter.bf = c(5*57, 
               5*43 + 4*(51 - 43) + 3*(53 - 51) + 2*(59 - 53) + 1*(60 - 59),
               5*60, 
               5*44 + 4*(46 - 44) + 3*(49 - 46) + 2*(62 - 49)),
  obj = c(-1324705.92,
          -1324701.57,
          -1326411.67,
          -1326413.04),
  total.fit.time = c(7.1 + 64.3 + 123.8, 
                     2.2 + 47.8 + 66.5, 
                     5.5 + 16.0 + 30.8, 
                     1.3 + 12.6 + 22.6) / 60
)

# The initialization timing includes one greedy iteration.
res$time.per.init = res$time.per.init - res$time.per.iter.greedy

row.names(res) <- c("flashr.normal.mix", 
                    "flashier.normal.mix", 
                    "flashr.point.normal", 
                    "flashier.point.normal")
colnames(res) <- cols

options(knitr.kable.NA = "")
knitr::kable(res, 
             digits = 2, 
             col.names = cols,
             caption = "GTEx eQTL data")
```

## Montoro et al. 3' scRNA-seq data

Next, I fit the droplet-based 3' scRNA-seq dataset analyzed in Montoro et al. (2018) (the data can be obtained [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103354)). I removed all genes with nonzero counts in five or fewer cells, performed a log-plus-one transform of the data, and fit five FLASH factors using scale-mixture-of-normal and point-normal priors. 

The data matrix is 15k x 7k and takes up 838 MB of memory when loaded into `R` as a dense matrix. However, only 11.2% of entries are nonzero, so the data can also be loaded as a sparse `Matrix` object, in which case the data takes up 142 MB of memory (only `flashier` supports objects of class `Matrix`). All fits were performed on the `broadwl` partition of the `midway2` RCC cluster using 4 CPUs and 32 GB of memory.

```{r res2, echo = FALSE}
res <- data.frame(
  mem.VMS = c(11.69, 3.25, 0.96, 11.78, 3.17, 0.87),
  mem.RSS = c(11.44, 3.07, 0.78, 11.07, 2.98, 0.69),
  time.per.init = c(261.9, 23.5, 7.1, 253.7, 22.5, 6.2) / 5,
  time.per.iter.greedy = c(151.4 / (2 + 2 + 2 + 2 + 2),
                           c(33.1, 12.5) / (1 + 4 + 9 + 9 + 5),
                           239.1 / (4 + 2 + 2 + 6 + 4),
                           c(28.7, 7.6) / (1 + 4 + 9 + 9 + 5)),
  time.per.iter.bf = c(2990 / (5*63),
                       349.2 / (5*19 + 4*(46 - 19) + 2*(91 - 46)),
                       131.2 / (5*19 + 4*(46 - 19) + 2*(91 - 46) + 1*(92 - 91)),
                       943.6 / (5*21),
                       c(30.0, 7.7) / (5*2 + 4*(6 - 2) + 3*(7 - 6) + 1*(8 - 7))),
  niter.bf = c(5*63, 
               5*19 + 4*(46 - 19) + 2*(91 - 46),
               5*19 + 4*(46 - 19) + 2*(91 - 46) + 1*(92 - 91),
               5*21, 
               rep(5*2 + 4*(6 - 2) + 3*(7 - 6) + 1*(8 - 7), 2)),
  obj = c(2354585.01,
          2354637.37,
          2354649.51, 
          2322664.76, 
          rep(2319043.85, 2)),
  total.fit.time = c(261.9 + 151.4 + 2990, 
                     23.5 + 33.1 + 349.2, 
                     7.1 + 12.5 + 131.2, 
                     253.7 + 239.1 + 943.6, 
                     22.5 + 28.7 + 30.0,
                     6.2 + 7.6 + 7.7) / 60
)

res$time.per.init <- res$time.per.init - res$time.per.iter.greedy

row.names(res) <- c("flashr.normal.mix", 
                    "flashier.normal.mix.", 
                    "flashier.sprs.normal.mix",
                    "flashr.point.normal",
                    "flashier.point.normal",
                    "flashier.sprs.point.normal")
colnames(res) <- cols

options(knitr.kable.NA = "")
knitr::kable(res, 
             digits = 2, 
             col.names = cols,
             caption = "Montoro 3' scRNA data")
```

## Montoro et al. full-length scRNA-seq data

Finally, I fit the larger full-length scRNA "PulseSeq" dataset from Montoro et al. (2018). The dataset is about ten times larger than the droplet-based scRNA-seq dataset, so it was not feasible to use `flashr`. I again removed genes with nonzero counts in five or fewer cells, performed a log-plus-one transform of the data, and fit five factors using scale-mixture-of-normal and point-normal priors. 

The dataset is 19k x 66k, with 10.6% of entries not equal to zero, and occupies 1.5 GB of memory when loaded into `R` as a sparse `Matrix` object. (I did not attempt to fit a larger `matrix` object.) The fit was again performed on `broadwl` using 4 CPUs and 32 GB of memory.

```{r res3, echo = FALSE}
res <- data.frame(
  mem.VMS = c(7.69, 7.69),
  mem.RSS = c(7.54, 7.54),
  time.per.init = c(61.0, 58.0) / 5,
  time.per.iter.greedy = c(144.8 / (1 + 4 + 13 + 21 + 13), 
                            115.3 / (1 + 4 + 14 + 21 + 13)),
  time.per.iter.bf = c(53.7 / (5*2 + 4*(3 - 2) + 3*(4 - 3) + 2*(5 - 4) + 1*(6 - 5)), 
                       16.3 / (5*1 + 3*(2 - 1))),
  niter.bf = c(5*2 + 4*(3 - 2) + 3*(4 - 3) + 2*(5 - 4) + 1*(6 - 5), 
               5*1 + 3*(2 - 1)),
  obj = c(125716192.04, 125635634.77),
  total.fit.time = c(61 + 144.8 + 53.7,
                     58 + 115.3 + 16.3) / 60
)

res$time.per.init = res$time.per.init - res$time.per.iter.greedy

row.names(res) <- c("flashier.sprs.normal.mix", "flashier.sprs.point.normal")
colnames(res) <- cols

options(knitr.kable.NA = "")
knitr::kable(res, 
             digits = 2, 
             col.names = cols,
             caption = "Montoro full-length scRNA data")
```

## Code

The benchmarking code can be browsed [here](https://github.com/willwerscheid/FLASHvestigations/tree/master/code/flashier_bench). The `R` package versions used are those that appear in the session information below ("other attached packages").
